File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        se = set()
        l, r, result = 0, 0, 0
        while r < len(s):
            while s[r] in se:
                se.remove(s[l])
                l += 1
            se.add(s[r])
            result = max(result, len(se))
            r += 1
        return result

File: Data Structures & Algorithms/task-scheduling/submission-3.java
class Solution {
    public int leastInterval(char[] tasks, int n) {
        Map<Character, Integer> map = new HashMap<>();
        for (char c : tasks) {
            map.put(c, map.getOrDefault(c, 0) + 1);
        }

        // Max-heap to get the most frequent task first
        PriorityQueue<Integer> heap = new PriorityQueue<>((a, b) -> Integer.compare(b, a));
        heap.addAll(map.values());

        int result = 0;
        // Queue to store tasks during their cooldown period
        Queue<int[]> cooldown = new LinkedList<>();

        while (!heap.isEmpty() || !cooldown.isEmpty()) {
            result++;

            if (!heap.isEmpty()) {
                int currentTaskCount = heap.poll() - 1;
                if (currentTaskCount > 0) {
                    cooldown.add(new int[]{currentTaskCount, result + n});
                }
            }

            // If a task's cooldown has ended, add it back to the heap
            if (!cooldown.isEmpty() && cooldown.peek()[1] == result) {
                heap.add(cooldown.poll()[0]);
            }
        }

        return result;
    }
}
